clicommand/git_credentials_helper.go

```diff
@@ -0,0 +1,134 @@
+		DebugHTTPFlag,
+
+		// Global flags
+		NoColorFlag,
+		DebugFlag,
+		LogLevelFlag,
+		ExperimentsFlag,
+		ProfileFlag,
+	},
+	Action: func(c *cli.Context) error {
+		ctx := context.Background()
+		ctx, cfg, l, _, done := setupLoggerAndConfig[GitCredentialsHelperConfig](ctx, c)
+		defer done()
+
+		l.Info("Authenticating checkout using Buildkite Github App Credentials...")
+
+		// ie, if the flags are from the command line rather than from the environment, which is how they should be passed
+		// to this process when it's called through the job executor
+		if os.Getenv("BUILDKITE_JOB_ID") == "" {
+			l.Warn("📎💬 It looks like you're calling this command directly in a step, rather than having the agent automatically call it")
+			l.Warn("This command is intended to be used as a git credential helper, and not called directly.")
+		}
+
+		// git passes the details of the current clone process to the credential helper via stdin
+		// we need to parse this to get the repo URL
+		// see: https://git-scm.com/docs/git-credential
+		stdin, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			return fatal(c, l, fmt.Errorf("failed to read stdin: %v", err))
+		}
+
+		lines := strings.Split(string(stdin), "\n")
+
+		host := ""
+		path := ""
+		protocol := ""
+		for _, line := range lines {
+			if p, ok := strings.CutPrefix(line, "protocol="); ok {
+				protocol = strings.TrimSpace(p)
+			}
+			if p, ok := strings.CutPrefix(line, "host="); ok {
+				host = strings.TrimSpace(p)
+			}
+			if p, ok := strings.CutPrefix(line, "path="); ok {
+				path = strings.TrimSpace(p)
+			}
+		}
+
+		if protocol != "https" {
+			return fatal(c, l, fmt.Errorf("git remote must be using the https protocol to use Github App authentication"))
+		}
+
+		ep, err := url.JoinPath(host, path)
+		if err != nil {
+			return fatal(c, l, fmt.Errorf("failed to join host and path: %v", err))
+		}
+
+		repo := protocol + "://" + ep
+
+		client := api.NewClient(l, loadAPIClientConfig(cfg, "AgentAccessToken"))
+		tok, _, err := client.GetGithubCodeAccessToken(ctx, repo, cfg.JobID)
+		if err != nil {
+			return fatal(c, l, fmt.Errorf("failed to get github app credentials: %v", err))
+		}
+
+		fmt.Fprintln(c.App.Writer, "username=token")
+		fmt.Fprintln(c.App.Writer, "password="+tok)
+		fmt.Fprintln(c.App.Writer, "")
+
+		l.Info("Authentication successful! 🎉")
+
+		return nil
+	},
+}
+
+// fatal is a helper function that logs an error and outputs a dummy password
+// git continues with clones etc even when the credential helper fails, so we should output something that will 100% cause
+// the clone to fail
+// this function always returns a cli.ExitError
+func fatal(c *cli.Context, l logger.Logger, err error) error {
+	l.Error("Error: %v. Authentication will proceed, but will fail.", err)
+	fmt.Fprintln(c.App.Writer, "username=fail")
+	fmt.Fprintln(c.App.Writer, "password=fail")
+	fmt.Fprintln(c.App.Writer, "quit=1")
+	fmt.Fprintln(c.App.Writer, "")
+
+	return cli.NewExitError("", 1)
```

Comment:
as discussed on call, having the message be passed through `l.Error` rather than returned as part of the exitError makes the text red, which is preferable.